ghash/src/lib.rs

```diff
@@ -47,75 +23,125 @@ pub type Tag = universal_hash::Output<GHash>;
     type KeySize = U16;
 
     /// Initialize GHASH with the given `H` field element
     fn new(h: &Key) -> Self {
-        let mut h = *h;
-        h.reverse();
-
-        #[allow(unused_mut)]
-        let mut h_polyval = mulX_POLYVAL(&h);
-
-        #[cfg(feature = "zeroize")]
-        h.zeroize();
-
-        #[allow(clippy::let_and_return)]
-        let result = GHash(Polyval::new(&h_polyval));
-
-        #[cfg(feature = "zeroize")]
-        h_polyval.zeroize();
-
-        result
+        unsafe {
+            let bs_mask = _mm_set_epi64x(BS_MASK1, BS_MASK2);
+            // `_mm_loadu_si128` performs an unaligned load
+            #[allow(clippy::cast_ptr_alignment)]
+            let h = _mm_loadu_si128(h.as_ptr() as *const __m128i);
+            let h = _mm_shuffle_epi8(h, bs_mask);
+            let y = _mm_setzero_si128();
+            Self { h, y }
+        }
     }
 }
 
+macro_rules! xor {
+    ($e1:expr, $e2:expr, $e3:expr, $e4:expr,) => {
+        _mm_xor_si128(_mm_xor_si128($e1, $e2), _mm_xor_si128($e3, $e4))
+    };
+    ($e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr,) => {
+        _mm_xor_si128(
+            $e1,
+            _mm_xor_si128(_mm_xor_si128($e2, $e3), _mm_xor_si128($e4, $e5)),
+        )
+    };
+}
+
 impl UniversalHash for GHash {
     type BlockSize = U16;
 
-    /// Input a field element `X` to be authenticated
+    #[inline]
     fn update(&mut self, x: &Block) {
-        let mut x = *x;
-        x.reverse();
-        self.0.update(&x);
+        unsafe {
+            let bs_mask = _mm_set_epi64x(BS_MASK1, BS_MASK2);
+
+            let h = self.h;
+            // `_mm_loadu_si128` performs an unaligned load
+            #[allow(clippy::cast_ptr_alignment)]
+            let x = _mm_loadu_si128(x.as_ptr() as *const __m128i);
+            let x = _mm_shuffle_epi8(x, bs_mask);
+            let y = _mm_xor_si128(self.y, x);
+
+            let h0 = h;
+            let h1 = _mm_shuffle_epi32(h, 0x0E);
+            let h2 = _mm_xor_si128(h0, h1);
+            let y0 = y;
+
+            // Multiply values partitioned to 64-bit parts
+            let y1 = _mm_shuffle_epi32(y, 0x0E);
+            let y2 = _mm_xor_si128(y0, y1);
+            let t0 = _mm_clmulepi64_si128(y0, h0, 0x00);
+            let t1 = _mm_clmulepi64_si128(y, h, 0x11);
+            let t2 = _mm_clmulepi64_si128(y2, h2, 0x00);
+            let t2 = _mm_xor_si128(t2, _mm_xor_si128(t0, t1));
+            let v0 = t0;
+            let v1 = _mm_xor_si128(_mm_shuffle_epi32(t0, 0x0E), t2);
+            let v2 = _mm_xor_si128(t1, _mm_shuffle_epi32(t2, 0x0E));
+            let v3 = _mm_shuffle_epi32(t1, 0x0E);
+
+            // Do the corrective 1-bit shift (255->256)
+            let v3 = _mm_or_si128(_mm_slli_epi64(v3, 1), _mm_srli_epi64(v2, 63));
+            let v2 = _mm_or_si128(_mm_slli_epi64(v2, 1), _mm_srli_epi64(v1, 63));
+            let v1 = _mm_or_si128(_mm_slli_epi64(v1, 1), _mm_srli_epi64(v0, 63));
+            let v0 = _mm_slli_epi64(v0, 1);
```

Comment:
Really this is the only part of the code where POLYVAL and GHASH differ (this shift is effectively replaced by `mulX_POLYVAL`).

Perhaps you could try updating the POLYVAL implementation and see if you still get equivalent performance? Then AES-GCM-SIV can benefit too.